### PR TITLE
Adding Training Model Code 

### LINKED ISSUE #12: Add Training and Predicting Code

### PR DESCRIPTION
**Pull request by AdityaSoni19031997**
Fix : #12

### CODE REVIEW

File: apis/train.py [hunk 1]
```diff
@@ -1,8 +1,69 @@
...
-        return {'success': 'Training pipeline successful'}
\ No newline at end of file
+        self.train()
+        return self.train() # {'success': 'Training pipeline successful',}
+
+    def train(self, file=None, target_var=None):
+
+        # curretnly it's having nothing as i/p
+        start_time = datetime.now()
+        MODELS_ROOT = "./apis/models" # will be changed init
```
Comment (shaz13) [inline apis/train.py]: Let's make dir at `./core/models` and store it there. apis directory is specified to namespaces and api methods only

File: apis/train.py [hunk 2]
```diff
@@ -1,8 +1,69 @@
...
+
+        if target_var is None:
+            response = "Outcome"
+        else:
+            response = target_var
+
+        if file is None: # sample file
+            df = pd.read_csv("./apis/diabetes.csv")
```
Comment (shaz13) [inline apis/train.py]: Same here. `/core/datasets`

File: apis/train.py (hunk 1 above)
Comment (AdityaSoni19031997) [reply apis/train.py]: okay

File: apis/train.py (hunk 2 above)
Comment (AdityaSoni19031997) [reply apis/train.py]: ok

File: apis/train.py [hunk 3]
```diff
@@ -1,8 +1,69 @@
...
+        path_to_model_file = "logreg.pkl"
+        self._persist_to_disk(clf, path_to_model_file)
+        validation_predictions = clf.predict_proba(valid[predictors])[:, 1]
+        score = roc_auc_score(valid[[response]], validation_predictions)
+
+        return {
+                'success': 'Training pipeline successful',
+                "msg" : f"Succesfully Trained in {datetime.now() - start_time}",
+                "roc" : f"{score}",
+            }
```
Comment (shaz13) [inline apis/train.py]: Return also a 200 response. Also, msg --> message and let's restructure this into 
```
return {
"message" :...
"metric" : "roc"
"score" : 0.70 (rounded)
}, 200
```

File: apis/train.py [hunk 4]
```diff
@@ -1,8 +1,69 @@
...
+
+    def _split_dataset(self, df, validation_percentage, seed):
+        
+        state = RandomState(seed)
+        validation_indexes = state.choice(df.index, int(len(df.index) * validation_percentage), replace=False)
```
Comment (shaz13) [inline apis/train.py]: Reindent it. Line too long. Will fail once we put CI and Pylint checks

File: apis/train.py [hunk 5]
```diff
@@ -1,8 +1,69 @@
...
+
+    def _persist_to_disk(self, classifier, path_to_file):
+
+        with open(path_to_file, "wb") as f:
+            pickle.dump(classifier, f)
+
+        if os.path.isfile(path_to_file):
+            return {"msg" : f"Successfully saved model at {path_to_file} :)"}
+        else:
+            return {"msg" : "Failed to save the model :("}
```
Comment (shaz13) [inline apis/train.py]: Add empty line here at EOF

**Review by shaz13 — CHANGES_REQUESTED** [review]
Looks good PR. It needs some structural and nit changes. PTAL :)

**Comment by shaz13** [comment]
@AdityaSoni19031997  Can you fetch the latest master and make another commit. Currently, there's seems to be a  conflict

File: apis/config.py [hunk 6]
```diff
@@ -0,0 +1,5 @@
+from pathlib import Path
+
+APP_ROOT    = Path("./core")
+MODELS_ROOT = APP_ROOT / "models" / "logreg.pkl"
+TEMP_CSV    = APP_ROOT / "diabetes.csv"
```
Comment (shaz13) [inline apis/config.py]: Add new line here

Comment (shaz13) [reply apis/config.py]: TEMP_CSV => DATA_CSV 
Also, why is it temp?

File: apis/train.py [hunk 7]
```diff
@@ -1,8 +1,76 @@
...
+        else:
+            response = target_var
+
+        if file is None: # sample file
+            df = pd.read_csv(TEMP_CSV)
+        else:
+            df = pd.read_csv(file)
+
+        train, valid = self._split_dataset(df, 0.20, 12345)
+        predictors   = df.columns[:-1]
```
Comment (shaz13) [inline apis/train.py]: Use code-reformatter. Or pretty fiery from VSCode. Uneven spacing :)

Comment (AdityaSoni19031997) [reply apis/train.py]: this was on purpose actually;

File: apis/train.py [hunk 8]
```diff
@@ -1,8 +1,76 @@
...
+        
+        clf = LogisticRegression(C=0.1, solver='lbfgs')
+        clf.fit(X=train[predictors], y=train[response])
+
+        self._persist_to_disk(clf, MODELS_ROOT)
```
Comment (shaz13) [inline apis/train.py]: RENAME = MODEL_ROOT

File: apis/train.py (hunk 3 above)
Comment (shaz13) [reply apis/train.py]: LGTM

File: apis/train.py [hunk 9]
```diff
@@ -1,8 +1,76 @@
...
+            pickle.dump(classifier, f)
+
+        if os.path.isfile(path_to_file):
+            return {
+                "msg" : f"Successfully saved model at {path_to_file} :)"
+                }
+        else:
+            return {
+                "msg" : "Failed to save the model :("
+                }
```
Comment (shaz13) [inline apis/train.py]: Rename msg=> message plz 👍

**Review by shaz13 — CHANGES_REQUESTED** [review]
Small minor changes. Rest of it looks good.

File: apis/train.py (hunk 7 above)
Comment (shaz13) [reply apis/train.py]: Linting will fail 😄

File: apis/config.py (hunk 6 above)
Comment (AdityaSoni19031997) [reply apis/config.py]: it's temp as of now because use needs to place his/her data_path there; there could be multiple files as well;

**Review by AdityaSoni19031997 — COMMENTED** [review]
re-check once;

**Review by shaz13 — CHANGES_REQUESTED** [review]
Fails to execute. Also, no pandas dep in requirements. 
-> FileNotFoundError: [Errno 2] No such file or directory: 'core/models/logreg.pkl'

**Review by shaz13 — CHANGES_REQUESTED** [review]
Made new build options. Kindly pass them and will review once build passes

**Review by shaz13 — APPROVED** [review]
LGTM 👍
Thanks a bunch, @AdityaSoni19031997

**Pull request merged**